### PR TITLE
Cleanup GCE service Accounts

### PR DESCRIPTION
**Pull request by b10n1k**
Implementing and extending the cleanup_all.

In brief the `parser_service_accoun_list`:
- Make a request to get the accosiated emails with `projects/suse-sle-qa`
- for each email a new request returns the endpoint to the service
- and then for each of the accounts perform deletion

`parser_service_account_list` needs to  filter out the irrellative accounts
`get_vault_name` is decorated to return the accounts where their keys have extended the 24 hours

Signed-off-by: ybonatakis <ybonatakis@suse.com>

### CODE REVIEW

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:216]:
```
credentials = service_account.Credentials.from_service_account_info(
            self.__credentials.getPrivateKeyData()
        )
        service = googleapiclient.discovery.build('iam', 'v1', credentials=credentials)
```

please drop this and just use https://github.com/SUSE/pcw/pull/125/files#diff-0dfc4095c73002d356e289d8edf872dcb8a6624ba9adc082c0c331aa9789bb9cR22 instead . you duplicating the code

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:215]: this comment applies to whole PR . please read https://realpython.com/python-string-formatting/ . in project we currently using `#2` . I am considering moving to `#3` some day but for sure against fallback to `#1`

No about this specific message . You are not cleaning any Vault keys here you cleaning service accounts in google cloud . this also refers to function name

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:265]: what is "vault email" ?

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:227]: can you elaborate benefit of using decorator pattern here ? AFAIK decorator should be used when you need to do same thing *in several classes not linked by inheritance* than you can save yourself from *code duplication* . Can not find anything related to this topic here . In a fact you just break single flow into two pieces and now it is hard to understand what actually going on with data ...

**Review by asmorodskyi — CHANGES_REQUESTED** [review]
1. tests are failing 
2. not correct use of decorator design pattern
3. confusion around functions/variable names all over PR ( we not suppose to do any cleanup in vault but cleanup of service accounts in gce )

**Review by asmorodskyi — CHANGES_REQUESTED** [review]
1. tests are failing 
2. inappropriate use of decorator design pattern 
3. confusion in functions / variable names ( we suppose to cleanup not vault but google cloud service accounts )

**Comment by asmorodskyi** [comment]
sorry for spam :) after posting first comment I got 500 from github so I decided to do it again ...

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py:216]: i did that in the very beginning and it was failing. i didnt overthink and i left it when i see that it was working. if i fix the rest i might take a look. but until then i am not going to change it as i am afraid that it will require additional changes. i have to investigate why it wasnt working and what changes it needs

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py:227]: to be honest, i dont know how to answer this question and i try to avoid them in general (many reason out of the scope here). Everyone should do their research. for instance decorators do not apply _in several classes_ only. But i will tell you that. some small portion of my thinking is that `filterAfterTime` will apply to ec2 and azure later. Unfortunately i dont have idea how it will look like there so i do not provide the one-off solution. i expect that it will require some changes when i will have clear picture from the other providers.

> In a fact you just break single flow into two pieces and now it is hard to understand what actually going on with data ...

Thats not true. in fact this design is very common in many frameworks (django, Spring, etc). the name of the functions is also self-explanatory. The `get_vault_name` requests and returns the `['keys']` and then you want to `filterAfterTime`, to get the _old_ ones.  `get_vault_name` will never have to change. if believe that if you see what `filterAfterTime` does you will notice that import different packages and does quite different operation, which do not belong inside `get_vault_name` as you would expect from such function.

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py:265]: the take out from the `service_account['email']`

File: ocw/lib/gce.py
Comment (asmorodskyi) [reply ocw/lib/gce.py:216]: sure thing it is up to you to decide the order , just saying that we will need to revisit this before merging in any case or find really good reasons to keep this duplication

File: ocw/lib/gce.py
Comment (asmorodskyi) [reply ocw/lib/gce.py:265]: service account is entity from google cloud `Credentials` , right ? how it is related to vault ?

File: ocw/lib/gce.py
Comment (asmorodskyi) [reply ocw/lib/gce.py:227]: > to be honest, i dont know how to answer this question and i try to avoid them in general (many reason out of the scope here)

sorry I didn't get this part 

> some small portion of my thinking is that filterAfterTime will apply to ec2 and azure later.

currently decorators are not used in this project at all . this PR is about adding google service account cleanup . I would prefer that we will concentrate on this without doing several things at the same time . I am fine to add decorators to the project but I can not accept having this pattern in one single random place while whole project not using same approach . Please solve : 
> Unfortunately i dont have idea how it will look like 

and provide " one-off solution" in separate PR.

File: ocw/lib/decorators.py
Comment (asmorodskyi) [inline ocw/lib/decorators.py:10]: I assume `validAfterTime` is field from Google service account field ?

**Comment by asmorodskyi** [comment]
just to sum up : 
- cleanup_vault_keys
-  vault_email
- vault_keys
- delete_vault_keys 
- get_vault_name
- vault_account_tuple
- delete_vault_keys
all this variable/function names have nothing to do with Vault so their names are confusing . 
Just to give you a hint . In `delete_vault_keys`  [you are deleting](https://github.com/SUSE/pcw/pull/125/files#diff-0dfc4095c73002d356e289d8edf872dcb8a6624ba9adc082c0c331aa9789bb9cR316) using call to `googleapiclient.discovery.Resource.projects().serviceAccounts().keys().delete` which clearly saying that you actually deleting google *service account* keys which has zero linkage to Vault

File: ocw/lib/gce.py
Comment (cfconrad) [inline ocw/lib/gce.py:317]:
```suggestion
                            name=vault_key).execute()
```

File: ocw/lib/gce.py
Comment (cfconrad) [inline ocw/lib/gce.py:280]: I think we can give this a general name, as we just call delete key and nothging special. The special things regarding vault are on the `collect` side.
```suggestion
    def delete_keys(self, iam_service, keys):
```

File: ocw/lib/gce.py
Comment (cfconrad) [inline ocw/lib/gce.py:248]: Is this decorator configurable by the caller of `paresr_service_account_list()` ?

File: ocw/lib/decorators.py
Comment (cfconrad) [inline ocw/lib/decorators.py:15]: Maybe `filter_colllection_starts_with` and isn't it the same as https://docs.python.org/3/library/functions.html#filter

File: ocw/lib/decorators.py
Comment (cfconrad) [reply ocw/lib/decorators.py:10]: yeah for such common function, I think a you would need to introduce a common class object or just do it in `gce.py`

**Review by cfconrad — CHANGES_REQUESTED** [review]
Nice that you introduced some more python code styles, but I think we should write this code as easy as possible.
 1) collect all emails `l = parser_service_account_list()`
 2) filter list for deleteable accounts (regex read from config) `l = deletable_accounts(l)`
 3) collect all keys `k = get_vault_name(l)`
 4) filter old keys (age read from config) `k_old = only_old_keys(k)
 5) delete all keys `delete_vault_keys(k_old);
 6) filter all service accounts older then XXX (age read from config) `old_l = old_accounts(l)`
 7) Delete all old service accounts `delete_service_accounts(old_l)`

1-5 is what you did, just extract the vault specific in functions and I would read the parameters from the config file

6+7 are there, to cleanup the services accounts it self. I didn't saw it, maybe I missed it.

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: *this change is ready for review*

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: *changes applied and waiting for feedback*

File: ocw/lib/decorators.py
Comment (b10n1k) [reply ocw/lib/decorators.py]: *changes applied and waiting for feedback*

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: *this change is ready for review*

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: *this change is ready for review*

**Comment by b10n1k** [comment]
I have changed names in the last commit. i also replace a function with `filter_email_service_accounts` which also removes one of the decorators. 
I gave up of the _credentials_ issue as for now. @asmorodskyi had a good argument but that might need additional changes overall. A workaround could be to destroy the credentials as soon as the cleanup ends or created it with really short lifetime. 

Now the `delete_vaultopenqa_service_accounts` deletes accounts. However i am struggling to verify the operation. Even manual. 

let me know if i am in the right direction.

File: ocw/lib/decorators.py
Comment (b10n1k) [reply ocw/lib/decorators.py:1]: i would say yes but from POV is just another way to choice of implementation. the decorator has different 'scope'. one is a design pattern, the other is python build-in function. So _same_ is relative

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py:244]: it can become obviously, but i am not sure what do you have in mind exactly.

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: i havent started the writing of the testing yet. i am going to do it as a very last step. unless..? Do you find good reason to start with it now?

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:220]:
```suggestion
    def cleanup_serviceaccounts(self):
```

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:34]:
```
def iam_client(self):
        if self.__credentials.isExpired():
            self.__credentials.renew()
            self.__compute_client = None
        self.__project = self.__credentials.getPrivateKeyData()["project_id"]
        if self.__iam_client is None:
            credentials = service_account.Credentials.from_service_account_info(self.__credentials.getPrivateKeyData()
            )
            self.__iam_client = googleapiclient.discovery.build(
                "iam", "v1", credentials=credentials, cache_discovery=False
            )
        return self.__iam_client
```

this is expected version . Your approach  will lead to some problems which we already had in PCW and which cause a lot issues with stalled credentials and use  already expired credentials. 

BTW `iam_client` and `compute_client` are good candidates for use of your favorite decorator pattern . this can be done as next step

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:25]: this is always best practice and also here it is quite necessary AFAIK to call parent's init as very first thing to do

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:15]: if you will check `azure` and `ec2` you will see that they also have `__new__` and `__init__` and I can tell you that before they were the same as `GCE` class . Actually I kinda surprised that GCE still use old stile . So generally +1 to this change . But you did it a bit differently than it is done in `azure` / `ec2` . I can not find any issue with THAT approach , but I would like to discuss this with you ...

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:221]:
```suggestion
        self.log_info(f"## GCE Service account cleaning - {self.__project} ##")
```

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:272]:
```suggestion
    def get_service_accounts(self, iam_service):
```

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:301]: can you elaborate why collecting everything in simple array but returning set wrapped in tuple ?

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:225]:
```suggestion
        service_accounts_filtered_by_email = self.parser_service_account_list(service)
```

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:230]:
```suggestion
    def filter_by_time(self, service, service_account_list, time_limit=24):
```

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:303]:
```suggestion
    def delete_service_accounts(self, iam_service, vault_accounts):
```

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:329]:
```suggestion
            self.log_debug("Nothing to delete")
```

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:341]:
```suggestion
                            name='projects/-/serviceAccounts/{}'.format(account_email)).execute()
```

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:334]:
```suggestion
                        "Deletion of vault Service Account {} skipped due to dry run mode".format(account_email))
```

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:343]:
```suggestion
                        self.log_err("Fail to delete Service Account {} \n{}".format(account_email,err))
```

File: ocw/lib/gce.py
Comment (asmorodskyi) [inline ocw/lib/gce.py:300]: since very next line after this break is return anyway we can simply do this 
```suggestion
                return service_accounts_emails
```

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: *this change is ready for review*

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py:35]: *this change is ready for review*

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: *this change is ready for review*

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: *this change is ready for review*

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: *this change is ready for review*

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: *this change is ready for review*

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: *this change is ready for review*

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: *this change is ready for review*

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: *this change is ready for review*

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: *this change is ready for review*

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: *this change is ready for review*

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: *this change is ready for review*

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py]: not a big thing or difference. i thought it was the appropriate data structure because:
- immutability
- iteration is faster
- less memory consumption

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py:344]: IMO the whole log_* has wrong application here. i was so tempted to changed it but it is far from the scope of this task.
_debug_ has higher level than _info_. meaning that if the logging was used properly and the logging level was _info_ the debug messages would not showed up. Something that in this particular case it would be wrong as you want to have the information in the regular logs.

File: ocw/lib/gce.py
Comment (b10n1k) [reply ocw/lib/gce.py:15]: i will take a look on the other modules. As for now i have no idea.

Comment (asmorodskyi) [reply ocw/lib/gce.py:15]: ok

File: tests/test_gce.py
Comment (asmorodskyi) [inline tests/test_gce.py:68]: leftover ?

Comment (b10n1k) [reply tests/test_gce.py]: yes. removed. *ready for review*

**Review by asmorodskyi — APPROVED** [review]
LGTM

**Comment by b10n1k** [comment]
@asmorodskyi i added a small documentation in REAME. can you take another look?

**Review by asmorodskyi — CHANGES_REQUESTED** [review]
I don't think that we need such info in PCW README . I think this file should describe only things specific to PCW . And this is purely documentation of Google Cloud API . For Google we also deleting instances . For Azure we deleting resource groups , for AWS we deleting images , disk , snapshots , VPC's , networks . Do you really thing that PCW documentation is the right place to describe how to deal with all this types , how to authenticate with AWS and Azure ?

**Comment by b10n1k** [comment]
> I don't think that we need such info in PCW README . I think this file should describe only things specific to PCW . And this is purely documentation of Google Cloud API . For Google we also deleting instances . For Azure we deleting resource groups , for AWS we deleting images , disk , snapshots , VPC's , networks . Do you really thing that PCW documentation is the right place to describe how to deal with all this types , how to authenticate with AWS and Azure ?

My opinion is that this is a good place because it has the cleanup and it can have the relative information in one place. If not in README maybe in the wiki. But i can remove this from the README now and take a look later

**Comment by asmorodskyi** [comment]
Due to changes happened outside of pcw ( drop of vault ) this functionality is not needed anymore hence PR will be closed .

**Pull request closed without merging**
